chalk-engine/src/logic.rs

```diff
@@ -251,8 +251,18 @@ impl<I: Interner, C: Context<I>> Forest<I, C> {
     ) -> Table<I> {
         let mut table = Table::new(goal.clone(), context.is_coinductive(&goal));
         let (mut infer, subst, environment, goal) = context.instantiate_ucanonical_goal(&goal);
-        match goal.data(context.interner()) {
-            GoalData::DomainGoal(domain_goal) => {
+        let goal_data = goal.data(context.interner());
+
+        let is_outlives_goal = |dg: &DomainGoal<I>| {
```

Comment:
Hmm. Interesting. I think this is more evidence that the way we're currently handling `LifetimeOutlives` isn't ideal. But this is fine for now.